File: luasrc/model/cbi/v2ray/policy.lua
-- Copyright 2019 Xingwang Liao <kuoruan@gmail.com>
-- Licensed to the public under the MIT License.

local dsp = require "luci.dispatcher"
local uci  = require "luci.model.uci".cursor()

local m, s1, s2, o

local level_table = {}

uci:foreach("v2ray", "policy_level", function(s)
	if s.level then
		level_table[s[".name"]] = s.level
	end
end)

m = Map("v2ray", "%s - %s" % { translate("V2Ray"), translate("Policy") })

s1 = m:section(NamedSection, "main_policy", "policy")
s1.anonymous = true
s1.addremove = false

o = s1:option(Flag, "enabled", translate("Enabled"))
o.rmempty = false

o = s1:option(MultiValue, "levels", translate("Levels"), translate("Select policy levels"))
for k, v in pairs(level_table) do
	o:value(k, v)
end

o = s1:option(Flag, "system_stats_inbound_uplink", "%s - %s" % { translate("System"), translate("Stats inbound uplink") })
o.enabled  = "true"
o.disabled = "false"

o = s1:option(Flag, "system_stats_inbound_downlink", "%s - %s" % { translate("System"), translate("Stats inbound downlink") })
o.enabled  = "true"
o.disabled = "false"

s2 = m:section(TypedSection, "policy_level", translate("Policy Level"), translate("Add policy levels here"))
s2.anonymous = true
s2.addremove = true
s2.template = "cbi/tblsection"
s2.extedit = dsp.build_url("admin/services/v2ray/policy/levels/%s")
s2.create = function (...)
	local sid = TypedSection.create(...)
	if sid then
		m.uci:save("v2ray")
		luci.http.redirect(s2.extedit % sid)
		return
	end
end

o = s2:option(DummyValue, "level", translate("Level"))
o.cfgvalue = function (...)
	return Value.cfgvalue(...) or "?"
end

o = s2:option(DummyValue, "handshake", translate("Handshake"))
o.cfgvalue = function (...)
	return Value.cfgvalue(...) or "?"
end

o = s2:option(DummyValue, "conn_idle", translate("Connection Idle"))
o.cfgvalue = function (...)
	return Value.cfgvalue(...) or "?"
end

o = s2:option(DummyValue, "buffer_size", translate("Buffer Size"))
o.cfgvalue = function (...)
	return Value.cfgvalue(...) or "?"
end

return m

File: src/view/v2ray/policy.ts
"use strict";

"require form";
"require uci";
"require v2ray";
// "require view";

// @ts-ignore
return L.view.extend({
  load: function () {
    return v2ray.getSections("policy_level", "level");
  },
  render: function (policyLevels: ReturnType<typeof v2ray.getSections> = []) {
    const m = new form.Map(
      "v2ray",
      "%s - %s".format(_("V2Ray"), _("Policy")),
      _("Details: %s").format(
        '<a href="https://www.v2ray.com/en/configuration/policy.html#policyobject" target="_blank">PolicyObject</a>'
      )
    );

    const s1 = m.section(form.NamedSection, "main_policy", "policy");
    s1.anonymous = true;
    s1.addremove = false;

    let o;
    o = s1.option(form.Flag, "enabled", _("Enabled"));
    o.rmempty = false;

    o = s1.option(
      form.MultiValue,
      "levels",
      _("Levels"),
      _("Select policy levels")
    );
    for (const s of policyLevels) {
      o.value(s.value, s.caption);
    }

    o = s1.option(
      form.Flag,
      "system_stats_inbound_uplink",
      "%s - %s".format(_("System"), _("Stats inbound uplink"))
    );

    o = s1.option(
      form.Flag,
      "system_stats_inbound_downlink",
      "%s - %s".format(_("System"), _("Stats inbound downlink"))
    );

    const s2 = m.section(
      form.GridSection,
      "policy_level",
      _("Policy Level"),
      _("Add policy levels here")
    );
    s2.anonymous = true;
    s2.addremove = true;
    s2.sortable = true;

    o = s2.option(form.Value, "level", _("Level"));
    o.rmempty = false;
    o.datatype = "uinteger";

    o = s2.option(form.Value, "handshake", _("Handshake"));
    o.datatype = "uinteger";
    o.placeholder = "4";

    o = s2.option(form.Value, "conn_idle", _("Connection idle"));
    o.datatype = "uinteger";
    o.placeholder = "300";

    o = s2.option(form.Value, "uplink_only", _("Uplink only"));
    o.modalonly = true;
    o.datatype = "uinteger";
    o.placeholder = "2";

    o = s2.option(form.Value, "downlink_only", _("Downlink only"));
    o.modalonly = true;
    o.datatype = "uinteger";
    o.placeholder = "5";

    o = s2.option(form.Flag, "stats_user_uplink", _("Stats user uplink"));
    o.modalonly = true;

    o = s2.option(form.Flag, "stats_user_downlink", _("Stats user downlink"));
    o.modalonly = true;

    o = s2.option(form.Value, "buffer_size", _("Buffer size"));
    o.datatype = "uinteger";

    return m.render();
  },
});
